### ISSUE TITLE
reproduciblity compared with vanilla pytorch

### ISSUE THREAD

**Issue by annongitter** [PyTorchLightning/pytorch-lightning#4674]
## 🐛 Bug

I am trying to compare a vanilla pytorch code which I refactored as a part of my learning pytorch-lightning but I see that the `training_step` is iterating through a different order of my dataset inspite of me setting `seed_everything` and `deterministic=True` and `benchmark=False`

## Please reproduce using [the BoringModel and post here]
 I have this as the base, common for both vanilla mode and pytorch-lightning mode:

```
        train_loader = torch.utils.data.DataLoader(\
        MyDataSet(cfg=cfg, mode="train"), \
                            batch_size=16, \
                            shuffle=True)
```

In lightning I do:

```
    def training_step(self, batch, batch_idx):
        x, y, paths = batch
        
        if Pathlib("check_input_path_pl").is_file():
            append_write = 'a+' # append if already exists
        else:
            append_write = 'w' # make a new file if not
        with open("check_input_path_pl", append_write) as f:
            f.write(str(batch_idx) + " ==> " +str(paths[0]) + "\n")

        y_hat = self(x)
        loss = self.loss_func(y_hat, y)
       return loss
```

in vanilla pytorch mode I do:

```
    for batch_idx, (data, target, paths) in enumerate(train_loader):
        data, target = data.to(device), target.to(device)
        if Pathlib("check_input_path").is_file():
            append_write = 'a+' # append if already exists
        else:
            append_write = 'w' # make a new file if not
        with open("check_input_path", append_write) as f:
            f.write(str(batch_idx) + " ==> " +str(paths[0]) + "\n")
```

but when I compare `check_input_path`  with `check_input_path_pl` I see that after the first epoch, the order of the input file paths is different :( 

I do:

```
seed_everything(111)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False
```

in both vanilla and lightning mode.

### Expected behavior

both should yield same order. To make a simple check, I iterate through the `train_loader` as soon as I create it in both pytl and pyt like so:

```
        for epoch in range(0, 2):
            for batch_idx, (data, target, paths) in enumerate(train_loader):
                if Pathlib("paths2load").is_file():
                    append_write = 'a+' # append if already exists
                else:
                    append_write = 'w' # make a new file if not
                with open("paths2load", append_write) as f:
                    f.write(str(batch_idx) + " ==> " +str(paths[0]) + "\n")
```

and in both pyt and pytl the files show the same order. yet, when training, they load in different order. What could the reason be?

### Environment

pytorch-lightning==1.0.6
torch==1.6.0
torchvision==0.7.0
Debian Linux
PS: both pyt and pytl are running with the same virtualenv

**Comment by Borda** [comment]
Hello, just to clarify, you say that your data loader and the one inside PL do not produce the same order of samples with both set `shuffle=True`, correct?

**Comment by annongitter** [comment]
@Borda exactly that.

**Comment by Borda** [comment]
ok, mind adding this kind of test to our parity testing?
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/benchmarks/test_parity.py

**Comment by annongitter** [comment]
@Borda I also found something else which I found a bit strange. If I seed_everything (in both pyt and pytl), and called a Dataset within which I print out `np.random.rand(2,2)` and `torch.rand(2,2)`, I see that the `numpy` random numbers in both pyt and pytl are the same, but not for the torch random numbers. So I manually again seed everything within the dataset (then I do get the exact same numbers in pyt and pytl), but it still iterates through a different order. Strange.

**Comment by annongitter** [comment]
> ok, mind adding this kind of test to our parity testing?
> https://github.com/PyTorchLightning/pytorch-lightning/blob/master/benchmarks/test_parity.py

I can, but can someone else confirm that they encounter the same problem please?

**Comment by annongitter** [comment]
for the sake of completeness, here is the dataset classes (in case there is something unusual in what I am doing):

```
class ImageFolderWithPaths(datasets.ImageFolder):
    # subclass to generate path names aside from data.
    def __getitem__(self, index):
        original_tuple = super(ImageFolderWithPaths, self).__getitem__(index)
        path = self.imgs[index][0]
        tuple_with_path = (original_tuple + (path,))
        return tuple_with_path

class MyDataSet(torch.utils.data.Dataset):
    def __init__(self, cfg=None, mode=None, transform=None):
        seed_everything(111)
        torch.backends.cudnn.deterministic = True
        torch.backends.cudnn.benchmark = False

        self.labels, self.file_paths=[],[]
        self.mode=mode
        if self.mode=="train":
            data_path = Pathlib(cfg.data_root).joinpath(cfg.train_str)
        elif self.mode=="test":
            data_path = Pathlib(cfg.data_root).joinpath(cfg.test_str)
        else:
            print("Mode  not recognised, exiting..")
            sys.exit()
        datasets = ImageFolderWithPaths(  root=data_path    )
        # print("datasets.classes: ",    datasets.classes)
        # print(np.random.rand(2,2))
        # print(torch.rand(2,2))
        # import sys
        # sys.exit()
        
        for idx, (sample, target, path) in enumerate(datasets):
            self.labels.append(  target  )
            self.file_paths.append(path)
        
        # write_list_to_file("./filepaths.txt", self.file_paths)
        # write_list_to_file("./labels.txt", self.labels)
        # sys.exit()

        self.train_data_transforms = A.Compose([
            ### some img transforms
        ])
        self.test_data_transforms = A.Compose([
            ### some img transforms
        ])


    def __len__(self):
        return len(self.file_paths)

    def __getitem__(self, idx):
        label = self.labels[idx]
        file_path = self.file_paths[idx]

        # Read an image with OpenCV
        image = cv2.imread(file_path)

        # By default OpenCV uses BGR color space for color images,
        # so we need to convert the image to RGB color space.
        image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
        if self.train_data_transforms:
            if self.mode=="train":
                augmented = self.train_data_transforms(image=image)
            else:
                augmented = self.test_data_transforms(image=image)
            image = augmented['image']
        return image, label, file_path
```

**Comment by adityak2920** [comment]
@annongitter Hi, I am also facing the same problem. But I have another use case, where I am using separate data loaders for both renders and query dataset. But I am unable to figure out how can I wrap both the dataloaders in pytorch-lightning because we only have the option of train, val and test dataloaders and I want two dataloaders for the test or val.

**Comment by rohitgr7** [comment]
@adityak2920 mind reproduce it with https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing??

**Comment by Zethson** [comment]
This issue is still relevant.
